lib/spack/spack/directives.py

```diff
@@ -767,6 +768,35 @@ def build_system(*values, **kwargs):
     )
 
 
+@directive()
+def disinherit(dict_name: str):
```

Comment:
I really like `disinherit` because the user *has* to know that the versions are inherited for this to be useful (i.e., you wouldn't use it to reset version declarations you made in the same class).  I don't see how this can't be tied to that bit of implementation.

Reply:
In my mind "disinherit" means we do something against OO inheritance.

Reply:
How about `drop()`?  See the top-level comment below.

On OO: We only use inheritance in Spack packages for sharing (some) metadata and for passing along the build mechanism.  It is inheritance, but the directives and packager-implemented methods are not the polymorphic part.

For example, we don't have functions that take an `Mpich` and expect to be able to do the same things with any subclass of `Mpich`. The main client code for packages is the solver and the installer... and they only require that the same metadata dictionaries are present and that the phases/do_install API works.  So I don't think we're violating polymorphism here, or at least we never really had it to begin with.

The other clients for packages are other packages that depend on them... but the only polymorphism you'd expect there is if a package in one repository has the same name as a package in a lower-precedence repo.  In that situation the client package *might* ask for one of the versions in the base package.  But if it can't find it that's not awful... the resolver just says the version doesn't exist... and that's something that happens in the normal course of maintaining things.

Reply:
> How about drop()? See the top-level comment below.

:+1:  "drop" sounds better to me.

Reply:
> So I don't think we're violating polymorphism here, or at least we never really had it to begin with.

Sure, I didn't mean to say that (and if it was the case, it wouldn't be solved by just renaming :slightly_smiling_face: ). What I wanted to say is that it "sounds like it has to do" with undoing something related to OO, when in fact it isn't related to that.